### PR TITLE
#202 Add rule to prevent final constructors in doctrine entities

### LINKED ISSUE #202: Implement a rule that prevents final constructors in Doctrine Entities
Hello there,

I'm having quite a hard time getting my head around this : [final methods in Doctrine Entities](https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/architecture.html#entities). There's [a PR](https://github.com/phpstan/phpstan-doctrine/pull/143) that added a check that Entities cannot be final (the class itself), but [a comment](https://github.com/phpstan/phpstan-doctrine/pull/143#issuecomment-698225313) from the author said he doesn't want to add checks for final methods.

However, from what I can see in my Doctrine proxied classes in var/cache in production, *all* methods of the class is inherited/overridden in proxies. So it just cannot work with a final method, whichever it is.

Why hasn't this check been added then ?

I'm curious about your insights. Thanks!
gnutix

### PR DESCRIPTION
**Pull request by LeoVie**
This fixes #202.

- Prevent Doctrine entities from containing final constructors
- At the same time, don't forbid non-constructor final methods in Doctrine entities

What do you think about that?

### CODE REVIEW

**Comment by LeoVie** [comment]
The failing checks don't seem to be related to the changes of this PR.

File: src/Rules/Doctrine/ORM/EntityConstructorNotFinalRule.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+			return [];
+		}
+
+		$classReflection = $scope->getClassReflection();
+
+		if ($classReflection === null) {
+			throw new ShouldNotHappenException();
+		}
+
+		return [sprintf(
```
Comment (ondrejmirtes) [inline src/Rules/Doctrine/ORM/EntityConstructorNotFinalRule.php]: Please use RuleErrorBuilder here.

File: src/Rules/Doctrine/ORM/EntityConstructorNotFinalRule.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	{
+		if ($node->name->name !== '__construct') {
+			return [];
+		}
+
+		if (!$node->isFinal()) {
+			return [];
+		}
+
+		$classReflection = $scope->getClassReflection();
```
Comment (ondrejmirtes) [inline src/Rules/Doctrine/ORM/EntityConstructorNotFinalRule.php:43]: You're reporting in all classes, not just entities.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also, the rule needs to be registered in rules.neon, but only for bleedingEdge.

Thanks.

File: src/Rules/Doctrine/ORM/EntityConstructorNotFinalRule.php (hunk 2 above)
Comment (LeoVie) [reply src/Rules/Doctrine/ORM/EntityConstructorNotFinalRule.php:43]: If I'm correct, this should work now only for entities. I've taken this code from `EntityNotFinalRule` for the detection, if the class is an entity

```php
if ($this->objectMetadataResolver->isTransient($classReflection->getName())) {
	return [];
}

$metadata = $this->objectMetadataResolver->getClassMetadata($classReflection->getName());
if ($metadata !== null && $metadata->isEmbeddedClass === true) {
	return [];
}
```

**Comment by LeoVie** [comment]
Hi, sorry for the delay :)
I've updated the PR according to your suggestions. I hope, everything is correct now.

File: rules.neon [hunk 3]
```diff
@@ -23,10 +23,13 @@ rules:
 	- PHPStan\Rules\Doctrine\ORM\DqlRule
 	- PHPStan\Rules\Doctrine\ORM\RepositoryMethodCallRule
 	- PHPStan\Rules\Doctrine\ORM\EntityNotFinalRule
+	- PHPStan\Rules\Doctrine\ORM\EntityConstructorNotFinalRule
```
Comment (ondrejmirtes) [inline rules.neon]: You need to remove it from the `rules` section.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: rules.neon (hunk 3 above)
Comment (LeoVie) [reply rules.neon]: Ah, okay. I removed it there. :)

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**